src/model/model_index.nit

```diff
@@ -0,0 +1,590 @@
+class VisibilityComparator
+	super ScoreComparator
+
+	redef fun compare(o1, o2) do return o2.mentity.visibility <=> o1.mentity.visibility
+end
+
+# Compare two matches by their name in lexicographic order
+#
+# Generally, for a same score, we want to put A before Z.
+class NameComparator
+	super ScoreComparator
+
+	redef fun compare(o1, o2) do return o1.mentity.name <=> o2.mentity.name
+end
+
+# Compare two matches by their name length
+class NameLengthComparator
+	super ScoreComparator
+
+	redef fun compare(o1, o2) do return o1.mentity.name.length <=> o2.mentity.name.length
+end
+
+# Compare two matches by their full_name in lexicographic order
+#
+# Generally, for a same score, we want to put A before Z.
+class FullNameComparator
+	super ScoreComparator
+
+	redef fun compare(o1, o2) do return o1.mentity.full_name <=> o2.mentity.full_name
+end
+
+# Compare two matches by their full name length
+class FullNameLengthComparator
+	super ScoreComparator
+
+	redef fun compare(o1, o2) do return o1.mentity.full_name.length <=> o2.mentity.full_name.length
+end
+
+redef class MEntity
+
+	# Compare MEntity class kind
+	private fun compare_mentity(mentity: MEntity): Int do return 0
+end
+
+redef class MPackage
+	redef fun compare_mentity(mentity) do
+		if mentity isa MPackage then return 0
+		return -1
+	end
+end
+
+redef class MGroup
+	redef fun compare_mentity(mentity) do
+		if mentity isa MPackage then return 1
+		if mentity isa MGroup then return 0
+		return -1
+	end
+end
+
+redef class MModule
+	redef fun compare_mentity(mentity) do
+		if mentity isa MPackage then return 1
+		if mentity isa MGroup then return 1
+		if mentity isa MModule then return 0
+		return -1
+	end
+end
+
+redef class MClass
+	redef fun compare_mentity(mentity) do
+		if mentity isa MPackage then return 1
+		if mentity isa MGroup then return 1
+		if mentity isa MModule then return 1
+		if mentity isa MClass then return 0
+		return -1
+	end
+end
+
+redef class MClassDef
+	redef fun compare_mentity(mentity) do
+		if mentity isa MPackage then return 1
+		if mentity isa MGroup then return 1
+		if mentity isa MModule then return 1
+		if mentity isa MClass then return 1
+		if mentity isa MClassDef then return 0
+		return -1
+	end
+end
+
+redef class MProperty
+	redef fun compare_mentity(mentity) do
+		if mentity isa MPackage then return 1
+		if mentity isa MGroup then return 1
+		if mentity isa MModule then return 1
+		if mentity isa MClass then return 1
+		if mentity isa MClassDef then return 1
+		if mentity isa MProperty then return 0
+		return -1
+	end
+end
+
+redef class MPropDef
+	redef fun compare_mentity(mentity) do
+		if mentity isa MPackage then return 1
+		if mentity isa MGroup then return 1
+		if mentity isa MModule then return 1
+		if mentity isa MClass then return 1
+		if mentity isa MClassDef then return 1
+		if mentity isa MProperty then return 1
+		if mentity isa MPropDef then return 0
+		return -1
+	end
+end
```

Comment:
`compare_mentity` is not that elegant. What about a slightly better:

~~~nit
class MEntityComparator
	super ScoreComparator
￼	redef fun compare(o1, o2) do return o1.mentity_kind_rank <=> o2.mentity_kind_rank
￼end

redef class MEntity
￼	fun mentity_kind_rank: Int abstract
end
redef class MPackage
￼	redef fun mentity_kind_rank do return 1
end
redef class MGroup
￼	redef fun mentity_kind_rank do return 2
end
# ...
~~~